File: src/skills/planning-with-files/scripts/session-catchup.py
#!/usr/bin/env python3
"""
Session Catchup Script for planning-with-files (oh-my-pi version)

Analyzes the previous session to find unsynced context after the last
planning file update. Designed to run on session_start.

Usage: python3 session-catchup.py [project-path]
"""

import json
import os
import sys
from pathlib import Path
from typing import Dict, List, Optional, Tuple

PLANNING_FILES = ["task_plan.md", "progress.md", "findings.md"]


def get_project_dir(project_path: str) -> Path:
    """Convert project path to oh-my-pi's storage path format."""
    # Normalize to an absolute path to ensure a stable representation
    # .as_posix() handles '\\' -> '/' conversion on Windows automatically
    resolved_str = Path(project_path).resolve().as_posix()

    # Sanitize path: replace separators with '-', remove ':' (Windows drives)
    sanitized = resolved_str.replace("/", "-").replace(":", "")

    # Apply naming convention: leading '-' and '_' -> '-'
    if not sanitized.startswith("-"):
        sanitized = "-" + sanitized
    sanitized_name = sanitized.replace("_", "-")

    # Check oh-my-pi data directory
    data_root_env = os.getenv("OMP_DATA_DIR")
    if data_root_env:
        data_root = Path(data_root_env)
    else:
        # Respect XDG_DATA_HOME if set, otherwise use default
        xdg_root = os.getenv("XDG_DATA_HOME")
        if xdg_root:
            data_root = Path(xdg_root) / "omp" / "storage"
        else:
            data_root = Path.home() / ".local" / "share" / "omp" / "storage"

    # oh-my-pi uses ~/.pi/agent/sessions/ for sessions
    omp_sessions = Path.home() / ".omp" / "agent" / "sessions"
    if omp_sessions.exists():
        return omp_sessions / sanitized_name

    # Fallback to data_root if ~/.omp doesn't exist
    return data_root / "session" / sanitized_name


def get_sessions_sorted(project_dir: Path) -> List[Path]:
    """Get all session files sorted by modification time (newest first)."""
    if not project_dir.exists():
        return []

    # oh-my-pi uses *.jsonl files
    sessions = list(project_dir.glob("*.jsonl"))
    # Filter out agent-specific sessions if any
    main_sessions = [s for s in sessions if not s.name.startswith("agent-")]
    return sorted(main_sessions, key=lambda p: p.stat().st_mtime, reverse=True)


def parse_session_messages(session_file: Path) -> List[Dict]:
    """Parse all messages from a session file, preserving order."""
    messages: List[Dict] = []

    # oh-my-pi uses JSONL format (one JSON object per line)
    with open(session_file, "r") as f:
        for line_num, line in enumerate(f):
            try:
                data = json.loads(line)
                if isinstance(data, dict):
                    data["_line_num"] = line_num
                    messages.append(data)
            except json.JSONDecodeError:
                # Ignore malformed lines to be resilient to partial writes
                pass

    return messages


def find_last_planning_update(messages: List[Dict]) -> Tuple[int, Optional[str]]:
    """
    Find the last time a planning file was written/edited.
    Returns (line_number, filename) or (-1, None) if not found.
    """
    last_update_line = -1
    last_update_file = None

    for msg in messages:
        msg_type = msg.get("type")

        # oh-my-pi uses 'message' type for all messages
        if msg_type == "message":
            msg_data = msg.get("message", {})
            content = msg_data.get("content", [])

            if isinstance(content, list):
                for item in content:
                    if item.get("type") == "tool_use":
                        tool_name = item.get("name", "")
                        tool_input = item.get("input", {})

                        # Check for Write/Edit tools
                        if tool_name in ("Write", "Edit"):
                            file_path = tool_input.get("file_path", "")
                            for pf in PLANNING_FILES:
                                if file_path.endswith(pf):
                                    last_update_line = msg["_line_num"]
                                    last_update_file = pf

        # Also check for tool_result type in oh-my-pi
        elif msg_type == "tool_result":
            tool_name = msg.get("toolName", "")
            if tool_name in ("Write", "Edit"):
                # Check original request for file path
                parent = msg.get("parentId")
                if parent:
                    # Find parent message
                    for m in messages:
                        if m.get("id") == parent:
                            content = m.get("message", {}).get("content", [])
                            if isinstance(content, list):
                                for item in content:
                                    if item.get("type") == "tool_use":
                                        file_path = item.get("input", {}).get(
                                            "file_path", ""
                                        )
                                        for pf in PLANNING_FILES:
                                            if file_path.endswith(pf):
                                                last_update_line = msg["_line_num"]
                                                last_update_file = pf

    return last_update_line, last_update_file


def extract_messages_after(messages: List[Dict], after_line: int) -> List[Dict]:
    """Extract conversation messages after a certain line number."""
    result = []

    for msg in messages:
        if msg.get("_line_num", -1) <= after_line:
            continue

        msg_type = msg.get("type")

        if msg_type == "message":
            msg_data = msg.get("message", {})
            role = msg_data.get("role", "")
            content = msg_data.get("content", [])

            if role == "user":
                text_content = ""
                if isinstance(content, list):
                    for item in content:
                        if item.get("type") == "text":
                            text_content = item.get("text", "")
                            break
                elif isinstance(content, str):
                    text_content = content

                if text_content and len(text_content) > 20:
                    result.append(
                        {
                            "role": "user",
                            "content": text_content,
                            "line": msg["_line_num"],
                        }
                    )

            elif role == "assistant":
                text_content = ""
                tool_uses = []

                if isinstance(content, list):
                    for item in content:
                        if item.get("type") == "text":
                            text_content = item.get("text", "")
                        elif item.get("type") == "tool_use":
                            tool_name = item.get("name", "")
                            tool_input = item.get("input", {})
                            if tool_name == "Edit":
                                tool_uses.append(
                                    f"Edit: {tool_input.get('file_path', 'unknown')}"
                                )
                            elif tool_name == "Write":
                                tool_uses.append(
                                    f"Write: {tool_input.get('file_path', 'unknown')}"
                                )
                            elif tool_name == "Bash":
                                cmd = tool_input.get("command", "")[:80]
                                tool_uses.append(f"Bash: {cmd}")
                            else:
                                tool_uses.append(f"{tool_name}")

                if text_content or tool_uses:
                    result.append(
                        {
                            "role": "assistant",
                            "content": text_content[:600] if text_content else "",
                            "tools": tool_uses,
                            "line": msg["_line_num"],
                        }
                    )

    return result


def main():
    project_path = sys.argv[1] if len(sys.argv) > 1 else os.getcwd()
    project_dir = get_project_dir(project_path)

    # Check if planning files exist (indicates active task)
    has_planning_files = any(Path(project_path, f).exists() for f in PLANNING_FILES)
    if not has_planning_files:
        # No planning files in this project; skip catchup to avoid noise
        return

    if not project_dir.exists():
        # No previous sessions, nothing to catch up on
        print(f"[planning-with-files] No session directory found at {project_dir}")
        return

    sessions = get_sessions_sorted(project_dir)
    if len(sessions) < 1:
        print(f"[planning-with-files] No session files found in {project_dir}")
        return

    # Find a substantial previous session
    target_session = None
    for session in sessions:
        if session.stat().st_size > 5000:
            target_session = session
            break

    if not target_session:
        target_session = sessions[0]  # Use most recent even if small

    messages = parse_session_messages(target_session)
    last_update_line, last_update_file = find_last_planning_update(messages)

    # No planning updates in the target session; skip catchup output.
    if last_update_line < 0:
        print(
            "[planning-with-files] No planning file updates found in previous session"
        )
        return

    # Only output if there's unsynced content
    messages_after = extract_messages_after(messages, last_update_line)

    if not messages_after:
        print("[planning-with-files] No unsynced messages found")
        return

    # Output catchup report
    print("\n[planning-with-files] SESSION CATCHUP DETECTED")
    print(f"Previous session: {target_session.stem}")

    print(f"Last planning update: {last_update_file} at message #{last_update_line}")
    print(f"Unsynced messages: {len(messages_after)}")

    print("\n--- UNSYNCED CONTEXT ---")
    for msg in messages_after[-15:]:  # Last 15 messages
        if msg["role"] == "user":
            print(f"USER: {msg['content'][:300]}")
        else:
            if msg.get("content"):
                print(f"ASSISTANT: {msg['content'][:300]}")
            if msg.get("tools"):
                print(f"  Tools: {', '.join(msg['tools'][:4])}")

    print("\n--- RECOMMENDED ---")
    print("1. Run: git diff --stat")
    print("2. Read: task_plan.md, progress.md, findings.md")
    print("3. Update planning files based on above context")
    print("4. Continue with task")


if __name__ == "__main__":
    main()

File: src/skills/planning-with-files/scripts/session-catchup.ts
#!/usr/bin/env tsx
/**
 * Session Catchup Script for planning-with-files (TypeScript version)
 *
 * Analyzes the previous session to find unsynced context after the last
 * planning file update. Can be run standalone or imported by the hook.
 *
 * Usage:
 *   Standalone: npx tsx session-catchup.ts [project-path]
 *   From hook:  import { runSessionCatchup } from './session-catchup'
 *
 * @version 2.0.0
 * @requires oh-my-pi >= 0.31.0
 */

import { existsSync, readFileSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { join, resolve } from "node:path";

const PLANNING_FILES = ["task_plan.md", "progress.md", "findings.md"];

interface SessionEntry {
	type: string;
	id?: string;
	parentId?: string;
	message?: {
		role?: string;
		content?: string | Array<{ type: string; text?: string }>;
	};
	[key: string]: unknown;
}

interface ExtractedMessage {
	role: "user" | "assistant";
	content: string;
	tools?: string[];
	line: number;
}

interface CatchupResult {
	hasUnsyncedContent: boolean;
	lastUpdateFile?: string;
	lastUpdateLine: number;
	unsyncedCount: number;
	messages: ExtractedMessage[];
	sessionFile?: string;
}

/**
 * Convert project path to oh-my-pi's storage path format
 */
function getProjectDir(projectPath: string): string {
	// Normalize to absolute path with forward slashes
	const resolvedPath = resolve(projectPath).replace(/\\/g, "/");

	// Sanitize path: replace separators with '-', remove ':' (Windows drives)
	let sanitized = resolvedPath.replace(/\//g, "-").replace(/:/g, "");

	// Ensure leading '-'
	if (!sanitized.startsWith("-")) {
		sanitized = `-${sanitized}`;
	}

	// oh-my-pi sessions directory
	const ompSessions = join(homedir(), ".omp", "agent", "sessions");
	if (existsSync(ompSessions)) {
		return join(ompSessions, sanitized);
	}

	// Fallback to legacy pi directory
	const piSessions = join(homedir(), ".pi", "agent", "sessions");
	if (existsSync(piSessions)) {
		return join(piSessions, sanitized);
	}

	// Ultimate fallback
	return join(homedir(), ".local", "share", "omp", "storage", "session", sanitized);
}

/**
 * Get all session files sorted by modification time (newest first)
 */
function getSessionsSorted(projectDir: string): string[] {
	if (!existsSync(projectDir)) {
		return [];
	}

	const { readdirSync } = require("node:fs");
	const { join } = require("node:path");

	try {
		const files = readdirSync(projectDir)
			.filter((f: string) => f.endsWith(".jsonl"))
			.filter((f: string) => !f.startsWith("agent-"))
			.map((f: string) => ({
				path: join(projectDir, f),
				mtime: statSync(join(projectDir, f)).mtimeMs,
			}))
			.sort((a: { mtime: number }, b: { mtime: number }) => b.mtime - a.mtime);

		return files.map((f: { path: string }) => f.path);
	} catch {
		return [];
	}
}

/**
 * Parse all messages from a session file
 */
function parseSessionMessages(sessionFile: string): SessionEntry[] {
	const messages: SessionEntry[] = [];

	try {
		const content = readFileSync(sessionFile, "utf-8");
		const lines = content.split("\n");

		for (let lineNum = 0; lineNum < lines.length; lineNum++) {
			const line = lines[lineNum].trim();
			if (!line) continue;

			try {
				const data = JSON.parse(line) as SessionEntry;
				messages.push(data);
			} catch {
				// Ignore malformed lines
			}
		}
	} catch (error) {
		console.error(`[planning-with-files] Error reading session file: ${error}`);
	}

	return messages;
}

/**
 * Find the last time a planning file was written/edited
 */
function findLastPlanningUpdate(messages: SessionEntry[]): {
	line: number;
	file?: string;
} {
	let lastUpdateLine = -1;
	let lastUpdateFile: string | undefined;

	for (let i = 0; i < messages.length; i++) {
		const msg = messages[i];
		const msgType = msg.type;

		// Check for message type with tool_use
		if (msgType === "message" && msg.message) {
			const content = msg.message.content;
			if (Array.isArray(content)) {
				for (const item of content) {
					if (item.type === "tool_use") {
						const toolName = item.name;
						const toolInput = item.input as { file_path?: string } | undefined;

						if ((toolName === "Write" || toolName === "Edit") && toolInput) {
							const filePath = toolInput.file_path || "";
							for (const pf of PLANNING_FILES) {
								if (filePath.endsWith(pf)) {
									lastUpdateLine = i;
									lastUpdateFile = pf;
								}
							}
						}
					}
				}
			}
		}
	}

	return { line: lastUpdateLine, file: lastUpdateFile };
}

/**
 * Extract conversation messages after a certain line
 */
function extractMessagesAfter(messages: SessionEntry[], afterLine: number): ExtractedMessage[] {
	const result: ExtractedMessage[] = [];

	for (let i = afterLine + 1; i < messages.length; i++) {
		const msg = messages[i];
		const msgType = msg.type;

		if (msgType === "message" && msg.message) {
			const role = msg.message.role;
			const content = msg.message.content;

			if (role === "user") {
				let textContent = "";
				if (Array.isArray(content)) {
					for (const item of content) {
						if (item.type === "text" && item.text) {
							textContent = item.text;
							break;
						}
					}
				} else if (typeof content === "string") {
					textContent = content;
				}

				if (textContent && textContent.length > 20) {
					result.push({
						role: "user",
						content: textContent,
						line: i,
					});
				}
			} else if (role === "assistant") {
				let textContent = "";
				const toolUses: string[] = [];

				if (Array.isArray(content)) {
					for (const item of content) {
						if (item.type === "text" && item.text) {
							textContent = item.text;
						} else if (item.type === "tool_use") {
							const toolName = item.name;
							const toolInput = item.input as { file_path?: string; command?: string } | undefined;
							if (toolName === "Edit" && toolInput) {
								toolUses.push(`Edit: ${toolInput.file_path || "unknown"}`);
							} else if (toolName === "Write" && toolInput) {
								toolUses.push(`Write: ${toolInput.file_path || "unknown"}`);
							} else if (toolName === "Bash" && toolInput) {
								const cmd = (toolInput.command || "").substring(0, 80);
								toolUses.push(`Bash: ${cmd}`);
							} else {
								toolUses.push(toolName);
							}
						}
					}
				}

				if (textContent || toolUses.length > 0) {
					result.push({
						role: "assistant",
						content: textContent.substring(0, 600),
						tools: toolUses,
						line: i,
					});
				}
			}
		}
	}

	return result;
}

/**
 * Run the session catchup analysis
 */
export function runSessionCatchup(projectPath: string): CatchupResult {
	const projectDir = getProjectDir(projectPath);

	// Check if planning files exist
	const hasPlanningFiles = PLANNING_FILES.some((f) => existsSync(join(projectPath, f)));

	if (!hasPlanningFiles) {
		return {
			hasUnsyncedContent: false,
			lastUpdateLine: -1,
			unsyncedCount: 0,
			messages: [],
		};
	}

	if (!existsSync(projectDir)) {
		return {
			hasUnsyncedContent: false,
			lastUpdateLine: -1,
			unsyncedCount: 0,
			messages: [],
		};
	}

	const sessions = getSessionsSorted(projectDir);
	if (sessions.length === 0) {
		return {
			hasUnsyncedContent: false,
			lastUpdateLine: -1,
			unsyncedCount: 0,
			messages: [],
		};
	}

	// Find a substantial previous session
	let targetSession: string | undefined;
	for (const session of sessions) {
		try {
			const stats = statSync(session);
			if (stats.size > 5000) {
				targetSession = session;
				break;
			}
		} catch {}
	}

	// If no substantial session, use the most recent
	if (!targetSession && sessions.length > 0) {
		targetSession = sessions[0];
	}

	if (!targetSession) {
		return {
			hasUnsyncedContent: false,
			lastUpdateLine: -1,
			unsyncedCount: 0,
			messages: [],
		};
	}

	const messages = parseSessionMessages(targetSession);
	const { line: lastUpdateLine, file: lastUpdateFile } = findLastPlanningUpdate(messages);

	if (lastUpdateLine < 0) {
		return {
			hasUnsyncedContent: false,
			lastUpdateLine: -1,
			unsyncedCount: 0,
			messages: [],
			sessionFile: targetSession,
		};
	}

	const messagesAfter = extractMessagesAfter(messages, lastUpdateLine);

	return {
		hasUnsyncedContent: messagesAfter.length > 0,
		lastUpdateFile,
		lastUpdateLine,
		unsyncedCount: messagesAfter.length,
		messages: messagesAfter,
		sessionFile: targetSession,
	};
}

/**
 * Format catchup result as human-readable output
 */
export function formatCatchupOutput(result: CatchupResult): string {
	if (!result.hasUnsyncedContent) {
		return "";
	}

	const lines: string[] = [];
	lines.push("");
	lines.push("[planning-with-files] SESSION CATCHUP DETECTED");

	if (result.sessionFile) {
		const sessionName = result.sessionFile.split("/").pop()?.replace(".jsonl", "");
		lines.push(`Previous session: ${sessionName}`);
	}

	lines.push(`Last planning update: ${result.lastUpdateFile} at message #${result.lastUpdateLine}`);
	lines.push(`Unsynced messages: ${result.unsyncedCount}`);
	lines.push("");
	lines.push("--- UNSYNCED CONTEXT ---");

	// Show last 15 messages
	const recentMessages = result.messages.slice(-15);
	for (const msg of recentMessages) {
		if (msg.role === "user") {
			lines.push(`USER: ${msg.content.substring(0, 300)}`);
		} else {
			if (msg.content) {
				lines.push(`ASSISTANT: ${msg.content.substring(0, 300)}`);
			}
			if (msg.tools && msg.tools.length > 0) {
				lines.push(`  Tools: ${msg.tools.slice(0, 4).join(", ")}`);
			}
		}
	}

	lines.push("");
	lines.push("--- RECOMMENDED ---");
	lines.push("1. Run: git diff --stat");
	lines.push("2. Read: task_plan.md, progress.md, findings.md");
	lines.push("3. Update planning files based on above context");
	lines.push("4. Continue with task");

	return lines.join("\n");
}

/**
 * Main function for CLI usage
 */
function main(): void {
	const projectPath = process.argv[2] || process.cwd();
	const result = runSessionCatchup(projectPath);

	if (result.hasUnsyncedContent) {
		console.log(formatCatchupOutput(result));
	}
}

// Run if called directly (not imported)
if (require.main === module) {
	main();
}
